### ISSUE TITLE
 approximate joint training？How does it work？

### ISSUE THREAD

**Issue by MenglaiWang** [rbgirshick/py-faster-rcnn#46]
How does joint training work？
I wonder how to coordinate the training between RPN and RCNN? Is it achieved by update the RPN with the RCNN fixed and then the opposite？ To be clearer, is the learning driven by RPN loss in one BackPropagation  and the next time by RCNN loss？？

**Comment by rbgirshick** [comment]
No, there is no alternation. Each mini-batch is used to compute a total loss that is the sum of four terms: RPN classification loss + RPN regression loss + Detection (R-CNN) classification loss + R-CNN bounding-box regression loss.

**Comment by MenglaiWang** [comment]
Thanks a lot !

**Comment by catsdogone** [comment]
The total loss doesn't equal to the sum of the four terms:
...
I0125 09:26:43.294986  5561 solver.cpp:571] Iteration 48420, lr = 5e-05
I0125 09:27:17.914181  5561 solver.cpp:242] Iteration 48440, loss = 0.891995
I0125 09:27:17.914214  5561 solver.cpp:258]     Train net output #0: loss_bbox = 0.204296 (* 1 = 0.204296 loss)
I0125 09:27:17.914221  5561 solver.cpp:258]     Train net output #1: loss_cls = 0.177461 (* 1 = 0.177461 loss)
I0125 09:27:17.914225  5561 solver.cpp:258]     Train net output #2: rpn_cls_loss = 0.0711229 (* 1 = 0.0711229 loss)
I0125 09:27:17.914229  5561 solver.cpp:258]     Train net output #3: rpn_loss_bbox = 0.126131 (* 1 = 0.126131 loss)
I0125 09:27:17.914234  5561 solver.cpp:571] Iteration 48440, lr = 5e-05
I0125 09:27:49.331830  5561 solver.cpp:242] Iteration 48460, loss = 5.59026
I0125 09:27:49.331864  5561 solver.cpp:258]     Train net output #0: loss_bbox = 0.0632553 (* 1 = 0.0632553 loss)
I0125 09:27:49.331871  5561 solver.cpp:258]     Train net output #1: loss_cls = 0.0835277 (* 1 = 0.0835277 loss)
I0125 09:27:49.331876  5561 solver.cpp:258]     Train net output #2: rpn_cls_loss = 0.00019457 (* 1 = 0.00019457 loss)
I0125 09:27:49.331881  5561 solver.cpp:258]     Train net output #3: rpn_loss_bbox = 0.00126853 (* 1 = 0.00126853 loss)
I0125 09:27:49.331886  5561 solver.cpp:571] Iteration 48460, lr = 5e-05
...
@MenglaiWang 
@rbgirshick

**Comment by makefile** [comment]
@catsdogone 
The loss value printed in log is smoothed loss which related to `average_loss` param. when you set average_loss: 1 in solver.proto, you will see that the total loss is equal to the sum of the four loss terms.
